File: backend/mongo/src/main/java/io/dkakunsi/bitapp/mongo/model/TransactionModel.java
package io.dkakunsi.bitapp.mongo.model;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;

import dev.morphia.annotations.Entity;
import io.dkakunsi.bitapp.common.EntityStatus;
import io.dkakunsi.bitapp.common.Id;
import io.dkakunsi.bitapp.transaction.entity.Transaction;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Entity("transactions")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class TransactionModel {

  @dev.morphia.annotations.Id
  private String id;
  private String userId;
  private String title;
  private String description;
  private LocalDate date;
  private LocalTime time;
  private String source;
  private String destination;
  private String loan;
  private Long amount;
  private String currency;
  private String category;
  private String type;

  private String status;
  private LocalDateTime createdAt;
  private LocalDateTime updatedAt;
  private String createdBy;
  private String updatedBy;

  public Transaction toTransaction() {
    var transactionId = Id.of(this.id);
    var userIdObj = Id.of(this.userId);

    return Transaction.builder()
        .id(transactionId)
        .user(userIdObj)
        .title(this.title)
        .description(this.description)
        .date(this.date)
        .time(this.time)
        .source(this.source != null ? Id.of(this.source) : null)
        .destination(this.destination != null ? Id.of(this.destination) : null)
        .loan(this.loan != null ? Id.of(this.loan) : null)
        .amount(this.amount)
        .currency(this.currency)
        .category(this.category != null ? Transaction.Category.valueOf(this.category) : null)
        .type(Transaction.Type.valueOf(this.type))
        .status(EntityStatus.valueOf(this.status))
        .createdAt(this.createdAt)
        .updatedAt(this.updatedAt)
        .createdBy(this.createdBy)
        .updatedBy(this.updatedBy)
        .build();
  }

  public static TransactionModel fromTransaction(Transaction transaction) {
    return new TransactionModel(
        transaction.id().value(),
        transaction.user().value(),
        transaction.title(),
        transaction.description(),
        transaction.date(),
        transaction.time(),
        transaction.source() != null ? transaction.source().value() : null,
        transaction.destination() != null ? transaction.destination().value() : null,
        transaction.loan() != null ? transaction.loan().value() : null,
        transaction.amount(),
        transaction.currency(),
        transaction.category() != null ? transaction.category().name() : null,
        transaction.type().name(),
        transaction.status().name(),
        transaction.createdAt(),
        transaction.updatedAt(),
        transaction.createdBy(),
        transaction.updatedBy());
  }
}

File: mobile/app/lib/features/transaction/data/transaction_model.dart
import 'dart:convert';

import 'package:bitapp/common/data/model/api_data.dart';
import 'package:bitapp/common/data/model/object_status.dart';
import 'package:bitapp/common/data/model/store_data.dart';
import 'package:bitapp/common/util/formatter.dart';
import 'package:bitapp/features/transaction/domain/transaction_category.dart';
import 'package:bitapp/features/transaction/domain/transaction_type.dart';
import 'package:flutter/material.dart';

class TransactionModel implements ApiData, StoreData {
  @override
  final String id;

  final String title;
  final double amount;
  final DateTime date;
  final String userId;
  final String? description;
  final TimeOfDay time;
  final TransactionType transactionType;
  final String? sourceAccountId;
  final String? destinationAccountId;
  final String? loanId;
  final ObjectStatus? status;
  final TransactionCategory? category;

  TransactionModel({
    required this.id,
    required this.userId,
    required this.title,
    required this.amount,
    required this.date,
    required this.time,
    required this.transactionType,
    required this.category,
    this.description,
    this.sourceAccountId,
    this.destinationAccountId,
    this.loanId,
    this.status,
  });

  @override
  String toRequestJson() {
    var json = {
      'id': id,
      'user': userId,
      'title': title,
      'description': description,
      'date': date.millisecondsSinceEpoch,
      'time': time.toInt(),
      'type': transactionType.value,
      'category': category?.value,
      'source': sourceAccountId,
      'destination': destinationAccountId,
      'amount': amount,
      'loan': loanId,
    };
    return jsonEncode(json);
  }

  @override
  Map<String, dynamic> toStoreJson() {
    return {
      'id': id,
      'user': userId,
      'title': title,
      'description': description,
      'date': date.millisecondsSinceEpoch,
      'time': time.toInt(),
      'type': transactionType.value,
      'category': category?.name,
      'sourceId': sourceAccountId,
      'destinationId': destinationAccountId,
      'amount': amount,
      'loanId': loanId,
      'status': status?.value,
    };
  }

  static List<TransactionModel> fromListResponsePayload(String response) {
    final List<dynamic> transactions = jsonDecode(response);
    return transactions.isNotEmpty
        ? transactions.map((e) => from(e)).toList()
        : [];
  }

  static TransactionModel fromResponsePayload(String response) {
    final data = jsonDecode(response);
    return from(data);
  }

  static TransactionModel from(dynamic data) => TransactionModel(
    id: data['id'],
    userId: data['user'],
    title: data['title'],
    description: data['description'],
    amount: (data['amount'] as num).toDouble(),
    date: DateTime.fromMillisecondsSinceEpoch(data['date']),
    time: TimeFormatter.fromInt(data['time'] as int),
    transactionType: TransactionType.valueOf(data['type']),
    category: TransactionCategory.valueOf(data['category']),
    sourceAccountId: data['sourceAccountId'],
    destinationAccountId: data['destinationAccountId'],
    loanId: data['loanId'],
    status:
        data['status'] != null
            ? ObjectStatus.valueOf(data['status'])
            : ObjectStatus.active,
  );
}
